### PR TITLE
Newsreader PDA App

### PR DESCRIPTION
**Pull request by DeityLink**
image

:cl:
* rscadd: Added the newsreader PDA app. Become addict to the media even in places with no newscasters! Only lets you read channels and Wanted issues.
* rscadd: The Librarian has the newsreader app installed by default on his PDA (also the Captain since they have all apps by default)
* tweak: Oh also the video game crash of 2525 has finally hit our sector and the price of PDA games has drastically fallen.

### CODE REVIEW

**Comment by DrSnips** [comment]
Looks like we need to drop spess R.O.B.

**Comment by GreatBigFailure** [comment]
I've been wanting this for ages, great job. A good news feed can really add flare to a shift.

**Comment by PrimeDSS13** [comment]
I like the PDA filter over the images, very nice.

**Comment by toomykins** [comment]
How do newscasters work exactly? would be a shame if this became a way to communicate with people out of radio range.

**Comment by DeityLink** [comment]
@toomykins the Newsreader app is read-only though. I'm not sure if newscasters rely on telecomms but then that'd be its own issue, that could be fixed in a later PR. It's not a viable alternative to proper radio anyway when it comes to two-way communications.

**Comment by DeityLink** [comment]
Which leads me to think that a portable newscaster would be a great Librarian-only item.

**Comment by Zth--** [comment]
can we make the PDA's background color a bit less horrendous?

**Comment by DeityLink** [comment]
Clearly not in this PR.

**Comment by ghost** [comment]
holy shit god bless you for adding the monochrome filter to photos on the PDA, I love it

File: code/game/machinery/newscaster.dm [hunk 1]
```diff
@@ -56,6 +58,27 @@
 	backup_author = ""
 	img = null
 	backup_img = null
+	img_pda = null
+	backup_img_pda = null
+
+/datum/feed_message/proc/make_PDA_img()
+	if (img)
+		img_pda = icon(img)
+		//turns the image grayscale then applies adds an olive coat of paint
```
Comment (DeityLink) [inline code/game/machinery/newscaster.dm]:
```suggestion
		//turns the image grayscale then applies an olive coat of paint
```

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 2]
```diff
@@ -328,10 +328,26 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+		qdel(A)
+	for(var/app_type in (typesof(/datum/pda_app) - /datum/pda_app))
+		var/datum/pda_app/app = new app_type()
+		app.onInstall(src)
+
 /obj/item/device/pda/heads/nt_supreme
 	name = "Nanotrasen Supreme Commander PDA"
 	ownjob = "Nanotrasen Supreme Commander"
 
+/obj/item/device/pda/heads/nt_supreme/New()
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm:339]: Don't copypaste these two New(), create a proc that achieves this. Also I don't understand why you are deleting all apps and then installing them again

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 3]
```diff
@@ -328,10 +328,26 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+	..()
+	for(var/A in applications)
+		qdel(A)
+	for(var/app_type in (typesof(/datum/pda_app) - /datum/pda_app))
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm]: Use `subtypesof`

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 4]
```diff
@@ -1120,6 +1142,57 @@ var/global/list/obj/item/device/pda/PDAs = list()
 							<a href='?src=\ref[cartridge.radio];bot=\ref[med];command=switch_power;user=\ref[usr]'>Turn [med.on ? "off" : "on"]</a> <br/>
 							</li>"}
 				dat += "</ul>"
+			if (PDA_APP_NEWSREADER)
+				var/datum/pda_app/newsreader/app = locate(/datum/pda_app/newsreader) in applications
+				switch(app.screen)
+					if (NEWSREADER_CHANNEL_LIST)
+						dat += {"<h4>Station Feed Channels</h4>"}
+						if(app)
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm]: You're inside `app.screen`, there's no way this check is necessary

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 5]
```diff
@@ -1120,6 +1142,57 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+							if(news_network.wanted_issue)
+								dat+= "<HR><b><A href='?src=\ref[src];choice=viewWanted'>Read Wanted Issue</A></b><HR>"
+							if( isemptylist(news_network.network_channels) )
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm]:
```suggestion
							if(isemptylist(news_network.network_channels))
```

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 4 above)
Comment (Zth--) [reply code/game/objects/items/devices/PDA/PDA.dm]: Actually the check should be before the switch

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 6]
```diff
@@ -1120,6 +1142,57 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+								dat+="<br><i>No active channels found...</i>"
+							else
+								for(var/datum/feed_channel/channel in news_network.network_channels)
+									if(channel.is_admin_channel)
+										dat+="<b><a href='?src=\ref[src];choice=readNews;channel=\ref[channel]'>[channel.channel_name]</a></b><br>"
+									else
+										dat+="<a href='?src=\ref[src];choice=readNews;channel=\ref[channel]'>[channel.channel_name]</a> [(channel.censored) ? ("***") : ""]<br>"
+					if (NEWSREADER_VIEW_CHANNEL)
+						if(app)
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm]: Same as above

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 7]
```diff
@@ -1120,6 +1142,57 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+										dat+="-[message.body] <br>"
+										if(message.img)
+											usr << browse_rsc(message.img_pda, "tmp_photo_pda[i].png")
+
+											dat+="<a href='?src=\ref[src];choice=showPhotoInfo;showPhotoInfo=\ref[message]'><img src='tmp_photo_pda[i].png' width = '192'></a><br>"
+										dat+="<font size=1>\[Story by <b>[message.author]</b>\]</font><HR>"
+
+							dat += {"<br><a href='?src=\ref[src];choice=viewChannels'>Back</a>"}
+					if (NEWSREADER_WANTED_SHOW)
+						if(app)
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm]: Same as above

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 8]
```diff
@@ -459,6 +475,12 @@ var/global/list/obj/item/device/pda/PDAs = list()
 	note = "Congratulations, your station has chosen the Thinktronic 5290 WGW-11 Series E-reader and Personal Data Assistant!"
 	silent = 1 //Quiet in the library!
 
+
+/obj/item/device/pda/librarian/New()
+	..()
+	var/datum/pda_app/newsreader/app = new /datum/pda_app/newsreader()
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/PDA.dm:469]:
```suggestion
	var/datum/pda_app/newsreader/app = new
```

File: code/game/objects/items/devices/PDA/apps.dm [hunk 9]
```diff
@@ -129,12 +130,35 @@ var/global/list/pda_app_menus = list(
...
+	menu = PDA_APP_NEWSREADER
+	icon = "pda_news"
+	var/datum/feed_channel/viewing_channel
+	var/screen = NEWSREADER_CHANNEL_LIST
+
+/datum/pda_app/newsreader/onInstall(var/obj/item/device/pda/device)
+	..()
+
+/datum/pda_app/newsreader/Destroy()
+	..()
```
Comment (Zth--) [inline code/game/objects/items/devices/PDA/apps.dm]: I don't think these two supercalls are necessary, they are implicit

File: code/game/machinery/newscaster.dm [hunk 10]
```diff
@@ -56,6 +58,27 @@
...
+/datum/feed_message/proc/make_PDA_img()
+	if (img)
+		img_pda = icon(img)
+		//turns the image grayscale then applies an olive coat of paint
+		img_pda.MapColors(rgb(77,77,77), rgb(150,150,150), rgb(28,28,28), rgb(128,128,0))
+		//lowers the brightness then ups the contrast so we get something that fits on a PDA screen
+		img_pda.MapColors(1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,-0.53,-0.53,-0.53,0)
+		img_pda.MapColors(1.75,0,0,0,0,1.75,0,0,0,0,1.75,0,0,0,0,1.75,-0.375,-0.375,-0.375,0)
+
+/datum/feed_message/proc/NewspaperCopy()//We only copy the vars we'll need
```
Comment (Zth--) [inline code/game/machinery/newscaster.dm:73]: The inconsistence between this proc name and the previous one really bothers me

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 8 above)
Comment (Zth--) [reply code/game/objects/items/devices/PDA/PDA.dm:469]: Well, you can dismiss this one, seems like all this file does it this way

File: code/game/machinery/newscaster.dm (hunk 10 above)
Comment (DeityLink) [reply code/game/machinery/newscaster.dm:73]: you mean CamelCase and underscore? or the actual naming?

File: code/game/objects/items/devices/PDA/apps.dm (hunk 9 above)
Comment (DeityLink) [reply code/game/objects/items/devices/PDA/apps.dm]: yep, forgot to clean them up. (I'm at work rn, I'll do the changes tonight)

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 2 above)
Comment (DeityLink) [reply code/game/objects/items/devices/PDA/PDA.dm:339]: That was a long time ago so I'm not 100% sure but there was either an issue of duplicate apps, or something about getting rid of the apps of the parent PDA.

**Comment by DeityLink** [comment]
Alright @Zth-- we good

File: code/game/machinery/newscaster.dm (hunk 10 above)
Comment (DeityLink) [reply code/game/machinery/newscaster.dm:73]: Changed make_PDA_image to ImagePDA

**Pull request merged**